test/rules/no-unnecessary-initializer/test.ts.lint

```diff
@@ -0,0 +1,23 @@
+let a: string | undefined = undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+var b: string | undefined = undefined;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+for (let c: number | undefined = undefined; c < 2; c++) {}
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+let d;
+
+const e = undefined;
+
+function f(x: string | undefined = undefined) {}
```

Comment:
It should not recommend making a parameter optional if it is followed by a required parameter:
```typescript
function f(foo: string|undefined = undefined, bar: number) {} // no error, parameter foo cannot be optional

function f2(foo: string|undefined = undefined, bar: number = 1) {} // foo can be optional
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [1]
```